Python/Product/TestAdapter.Analysis/TestAnalysisExtension.cs

```diff
@@ -162,15 +161,30 @@ public class Serialize {
         /// return those.  If there aren't any 'test*' tests return (if one at 
         /// all) the runTest overridden method
         /// </summary>
-        private static IEnumerable<KeyValuePair<string, IAnalysisSet>> GetTestCaseMembers(
+        private static IEnumerable<KeyValuePair<string, LocationInfo>> GetTestCaseMembers(
+            PythonAst ast,
+            string sourceFile,
             ModuleAnalysis analysis,
             AnalysisValue classValue
         ) {
+
+            IEnumerable<KeyValuePair<string, LocationInfo>> tests = null, runTest = null;
+            if (ast != null && !string.IsNullOrEmpty(sourceFile)) {
+                var walker = new TestMethodWalker(ast, sourceFile, classValue.Locations);
+                ast.Walk(walker);
+                tests = walker.Methods.Where(v => v.Key.StartsWith("test"));
+                runTest = walker.Methods.Where(v => v.Key.Equals("runTest"));
+            }
+
             var methodFunctions = classValue.GetAllMembers(analysis.InterpreterContext)
-                .Where(v => v.Value.Any(m => m.MemberType == PythonMemberType.Function || m.MemberType == PythonMemberType.Method));
+                .Where(v => v.Value.Any(m => m.MemberType == PythonMemberType.Function || m.MemberType == PythonMemberType.Method))
+                .Select(v => new KeyValuePair<string, LocationInfo>(v.Key, v.Value.SelectMany(av => av.Locations).FirstOrDefault(l => l != null)));
+
+            var analysisTests = methodFunctions.Where(v => v.Key.StartsWith("test"));
+            var analysisRunTest = methodFunctions.Where(v => v.Key.Equals("runTest"));
 
-            var tests = methodFunctions.Where(v => v.Key.StartsWith("test"));
-            var runTest = methodFunctions.Where(v => v.Key.Equals("runTest"));
+            tests = tests?.Union(analysisTests) ?? analysisTests;
```

Comment:
This will produce duplicates in some cases. Like when running TestDiscovererTests.DecoratedTests, which now fails because the expected sequence of 1 element now has 2 elements.

Reply:
Ah. Worked fine before I had to change it all to key value pairs here :)